src/__tests__/utils.js

```diff
@@ -0,0 +1,82 @@
+/* @flow */
+
+import Client from '../index';
+
+import type {
+  RequestFactory,
+  RequestPerformer,
+  ResponseParser,
+} from '../types';
+
+type TestRequest = {
+  uri: string,
+  method?: string,
+  body?: Object,
+};
+
+type JSON = Object | Array<JSON> | number | string;
+
+type TestResponse = {
+  status: number,
+  body?: JSON,
+};
+
+export class TestRequestFactory implements RequestFactory<TestRequest> {
+  createRequest = (uri: string, method?: string, body?: Object) => ({
+    uri,
+    method,
+    body,
+  });
+}
+
+export class TestRequestPerformer
+  implements RequestPerformer<TestRequest, TestResponse> {
+  static resolve = (request: TestRequest, response: TestResponse) =>
+    new TestRequestPerformer(() => Promise.resolve(response));
+
+  static reject = (request: TestRequest, response: TestResponse) =>
+    new TestRequestPerformer(() => Promise.reject(response));
+
+  constructor(resolver: (request: TestRequest) => Promise<TestResponse>) {
+    this.resolver = resolver;
+  }
+
+  resolver: (request: TestRequest) => Promise<TestResponse>;
+
+  performRequest = (request: TestRequest): Promise<TestResponse> =>
+    this.resolver(request);
+}
+
+export class TestResponseParser implements ResponseParser<TestResponse> {
+  static resolve = (response: TestResponse, result: *) =>
+    new TestRequestPerformer(() => Promise.resolve(result));
+
+  static reject = (response: TestResponse, result: *) =>
+    new TestRequestPerformer(() => Promise.reject(result));
+
+  constructor(resolver: (response: TestResponse) => Promise<*>) {
+    this.resolver = resolver;
+  }
+
+  resolver: (response: TestResponse) => Promise<*>;
+
+  parseResponse = (response: TestResponse): Promise<*> =>
+    this.resolver(response);
+}
+
+export const createTestClient = (
```

Comment:
I think understand high level whats going on, but maybe a little deeper dive might help.

Reply:
Let me know if you want to walk through this more! Glad to answer questions and make changes